jwk_test.go

```diff
@@ -17,40 +17,240 @@
+		},
+		D:      rsaTestKey.D,
+		Primes: []*big.Int{rsaTestKey.Primes[0]},
+	}
+	var jwk rawJsonWebKey
+	err := jwk.fromRsaPrivateKey(&brokenRsaPrivateKey)
+	if err != ErrUnsupportedKeyType {
+		t.Error("expected unsupported key type error, got", err)
+	}
+}
 
-	for _, invalid := range invalids {
+func TestRsaPrivateExcessPrimes(t *testing.T) {
+	brokenRsaPrivateKey := rsa.PrivateKey{
+		PublicKey: rsa.PublicKey{
+			N: rsaTestKey.N,
+			E: rsaTestKey.E,
+		},
+		D: rsaTestKey.D,
+		Primes: []*big.Int{
+			rsaTestKey.Primes[0],
+			rsaTestKey.Primes[1],
+			big.NewInt(3),
+		},
+	}
+	var jwk rawJsonWebKey
+	err := jwk.fromRsaPrivateKey(&brokenRsaPrivateKey)
+	if err != ErrUnsupportedKeyType {
+		t.Error("expected unsupported key type error, got", err)
+	}
+}
+
+func TestRoundtripEcPrivate(t *testing.T) {
+	for _, ecTestKey := range []*ecdsa.PrivateKey{ecTestKey256, ecTestKey384, ecTestKey521} {
 		var jwk rawJsonWebKey
-		err := json.Unmarshal([]byte(invalid), &jwk)
+		jwk.fromEcPrivateKey(ecTestKey)
+
+		ec2, err := jwk.ecPrivateKey()
+		if err != nil {
+			t.Error("problem converting ECDSA private -> JWK", err)
+		}
+
+		if !reflect.DeepEqual(ec2.Curve, ecTestKey.Curve) {
+			t.Error("ECDSA private curve mismatch")
+		}
+		if ec2.X.Cmp(ecTestKey.X) != 0 {
+			t.Error("ECDSA X mismatch")
+		}
+		if ec2.Y.Cmp(ecTestKey.Y) != 0 {
+			t.Error("ECDSA Y mismatch")
+		}
+		if ec2.D.Cmp(ecTestKey.D) != 0 {
+			t.Error("ECDSA D mismatch")
+		}
+	}
+}
+
+func TestMarshalUnmarshal(t *testing.T) {
+	kid := "DEADBEEF"
+
+	for _, key := range []interface{}{ecTestKey256, ecTestKey384, ecTestKey521, rsaTestKey} {
+		jwk := JsonWebKey{Key: key, KeyID: kid}
+		jsonbar, err := jwk.MarshalJSON()
+		if err != nil {
+			t.Error("problem marshaling", err)
+		}
+
+		var jwk2 JsonWebKey
+		err = jwk2.UnmarshalJSON(jsonbar)
+		if err != nil {
+			t.Error("problem unmarshalling", err)
+		}
+
+		jsonbar2, err := jwk2.MarshalJSON()
+		if err != nil {
+			t.Error("problem marshaling", err)
+		}
+
+		if !bytes.Equal(jsonbar, jsonbar2) {
+			t.Error("roundtrip should not lose information")
+		}
+
+		if jwk2.KeyID != kid {
+			t.Error("kid did not roundtrip JSON marshalling")
+		}
+	}
+}
+
+func TestMarshalUnmarshalInvalid(t *testing.T) {
+	keys := []interface{}{
+		// Empty keys
+		&rsa.PrivateKey{},
+		&ecdsa.PrivateKey{},
+		// Invalid keys
+		&ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				// Missing values in pub key
+				Curve: elliptic.P256(),
+			},
+		},
+		&ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				// Invalid curve
+				Curve: nil,
+				X:     ecTestKey256.X,
+				Y:     ecTestKey256.Y,
+			},
+		},
+		&ecdsa.PrivateKey{
+			// Valid pub key, but missing priv key values
+			PublicKey: ecTestKey256.PublicKey,
+		},
+		nil,
+	}
+
+	for _, key := range keys {
+		jwk := JsonWebKey{Key: key}
+		_, err := jwk.MarshalJSON()
 		if err == nil {
-			t.Error("parser incorrectly parsed invalid key", jwk)
+			t.Error("managed to serialize invalid key")
```

Comment:
For all the looped test cases, it would be helpful to output which case failed with the error message.